src/Migration.php

```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace P4\MasterTheme;
+
+/**
+ * A migration that can be recorded.
+ */
+abstract class Migration {
+	/**
+	 * Get a unique identifier, achieved here by using the FQCN.
+	 *
+	 * @return string The unique identifier.
+	 */
+	public static function get_id(): string {
+		return static::class;
+	}
+
+	/**
+	 * Perform the actual migration.
+	 */
+	abstract public static function run(): void;
```

Comment:
Can you give an example of a dependency a migration could have?

I think I made these static because in essence these migrations are just a function we want to run. That function has access to the same globally available things which are accessible everywhere when using WP (i.e. data from the db and settings in wp-config.php). So I'm not sure if there's anything left to inject as a dependency/parameter?

Reply:
I'm writing too much there so **TLDR; static leads to no isolation. No isolation leads to unstructured code. Unstructured code leads to pain.** <sub>It also leads to me ranting, I don't usually do this, sorry about that</sub>

> Can you give an example of a dependency a migration could have?  

I guess the shortcode converter could have been one ? Any tool that is useful to the current code and has to be used by the migration, to create a new state. In a perfect world, the migration in this PR would use a WP_Options dependency to get and update those (but of course it doesn't exist 😭). In a future patch, I could use the SplitTwoColumns class to run the existing method that convert old blocks on the fly, to patch all posts.

> That function has access to the same globally available things which are accessible everywhere when using WP  

Yes, exactly. Because we are using Wordpress, we tend to fall into the same flaws (or what I consider as flaws), and this is what I would love to not end up doing. **What I'm actually saying is, instead of using things the Wordpress way and calling everything from anywhere, let's make our code cleaner and tighter, and for the most part consider that the global state doesn't exist**. We obviously have to use it, but isolating our code will make it more resilient (because less interactions), less prone to global state unplanned changes (because no global), and more explicit on its needs and inputs/outputs (because typehinting all the things). (Also, more testable.)
My "beef" with static in general is that it's very often no better than using `global`, or the global state (and global state is hard to follow and often has no structure), and the global state is already completely overused by Wordpress. Of course if you push far enough, dependency injection is a container in the global state, and in the end all the code is in it, but having access to anything from anywhere leads your code to less structure and random undeclared coupling.  
 My beef with static methods here is that because it is static by default:
- no instantiation means you won't even think about passing an object as a function parameter, so you won't typehint what your migration runner actually runs, the scope is not obvious, the mental model is wide, I don't know who's using what and in what context.
- using static methods and global state doesn't invite you to declare your dependencies explicitly; the code is then less obvious, what the class and the methods are using is unknown until I read the full code.  
The smallest and more explicit the scope is, the easiest it is for me to reason about it. It's like reading a recipe with a list of all ingredients and tools at the beginning, I can almost get the taste of it without reading the rest, and I know right away what I'll need.
- testing code coupled to wordpress code means including all of wordpress, I'd love our code to be at least a bit more unit testable. Having a `WP_Options` object to inject or add as a parameter would allow us to test the migration. Having MigrationLog and MigrationList as parameters (as interfaces or mockable object) would allow us to easily test a migration run. Having a specific migration runner class with a specific command to run would allow developers to easily test their migration in isolation during development. I'm not even talking about writing tests, but thinking about making it testable usually leads to a cleaner code on the long term (at least for me).

Reply:
These are all valid remarks, and if you have control over the entire code base then there indeed should be no global state. However we cannot change the fact that WP puts certain things in the global scope. At best you can add a layer around these things to access them through an object, but that would only make them seem like they're encapsulated, while in the end any piece of code (plugins) can and will still interact with the values through the global that's being wrapped. Because of that those objects should be treated in the same way as though they were the global they're wrapping, so you're missing out on the main benefit of encapsulation while the code has become significantly more complex.

It's also a bit of a rare case here because of the goal of this class:

Specifically in this case the `Migration` class is intended to facilitate writing a script that will interact with the database. In WP interacting with the db can be done in multiple ways, and for convenience here it makes sense to allow them all, so that you can choose whether to use WP's API (`get_posts`, `update_post_meta`, ...) or write a raw SQL query, whichever makes more sense for the tasks the script needs to do. Actually it's not possible to disallow any of these functions anyway as they're globally available. It's a pity but that's just the API WP exposes.

So because we're dealing with scripts that already depend on using global functions to interact with the db, and so the contents of any migration script will already vary a lot, it helps to put the bare minimum inside of this class, and handle complexity in dedicated separate objects.

So, long story short, while in a lot of cases static methods provide less encapsulation, in this case it actually provides more, as we can provide separate objects and pass them as a parameter to the function (like [here](https://github.com/greenpeace/planet4-master-theme/pull/1172/files#diff-da4f24715e30c74e0a897d4de165372bR41)).